### ISSUE TITLE
colexec:

### ISSUE THREAD

**Issue by asubiotto** [cockroachdb/cockroach#49794]
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by asubiotto** [comment]
cc @rytaft

**Comment by asubiotto** [comment]
The coster choice could be made based on what the `vectorize` session setting is since all operators that can be vectorized will be if the setting is `on`, and wrapped otherwise.

**Comment by rytaft** [comment]
I like the idea of adding a separate coster for vectorization, with the possibility of some shared cost functions. `Coster` is already an interface, so it would be relatively straightforward to add another implementation.

**Comment by asubiotto** [comment]
Since we're going to remove the vectorize row count threshold (#53893), the only thing the optimizer needs to do is:
1) Create a cost model for the vectorized engine (how can we work together to do this?)
2) Depending on whether vectorized execution is set or not, use one cost model or another.
This removes the need for the optimizer to decide whether to use vectorized execution or not.

**Comment by rytaft** [comment]
Sorry I never responded to this. I think one thing we need to determine is whether (a) the logic in [coster.go](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/opt/xform/coster.go) remains the same and only the [parameters](https://github.com/cockroachdb/cockroach/blob/938d916aa400e5bb9f02d7e651ff9e5db766f7ee/pkg/sql/opt/xform/coster.go#L93) need to change, or (b) we need to completely rewrite much of the coster logic. [Coster](https://github.com/cockroachdb/cockroach/blob/938d916aa400e5bb9f02d7e651ff9e5db766f7ee/pkg/sql/opt/xform/coster.go#L54) is already an interface, so it wouldn't be too difficult to just create a new implementation if need be.

I think the majority of the work will require updating the [benchmarking tool](https://github.com/cockroachdb/cockroach/tree/master/pkg/sql/opt/opbench) that Justin wrote, and running benchmarks to determine the correct parameter settings for the new cost model.

Probably worth setting up a meeting to discuss the best way to share the work for this and make sure it's accounted for in the release planning. cc @RaduBerinde

**Comment by asubiotto** [comment]
> (a) the logic in coster.go remains the same and only the parameters need to change

My first reaction is that only the parameters would need to change, I wonder what @yuzefovich thinks about this.

> Probably worth setting up a meeting to discuss the best way to share the work for this and make sure it's accounted for in the release planning.

We currently have vectorized costing marked as out of scope for the 21.1 release but I think it'd be good to have a meeting regardless to define the work we need to do on either side.

**Comment by yuzefovich** [comment]
I remember @jordanlewis expressed a thought that there is no need for different cost models for two execution engines. The idea is that we will be using the vectorized engine in pretty much all of the cases once we remove `vectorize_row_count_threshold`, so we should simply adjust the parameters of the cost model to make it more aligned with vectorized costs (for example, lookup joins should become more expensive). The cases when the row-by-row engine ends up being used for some queries should be rare, so it is acceptable if the costing becomes less optimal for that engine.

That said, we will probably need to address the limitation of geospatial (Becca's prototype was in #53202) as well as remove a few cases from `colbuilder/execplan.go` in which we don't wrap the row-by-row processor (most notably `core.LocalPlanNode`).

**Comment by asubiotto** [comment]
> The cases when the row-by-row engine ends up being used for some queries should be rare, so it is acceptable if the costing becomes less optimal for that engine.

I don't think so. `vectorize = off` should still be a fully supported configuration for the time being and we shouldn't regress. There isn't much of a cost to having two cost models so it seems like an outsized negative impact for little gain.

**Comment by jordanlewis** [comment]
I think there's quite a big cost to having two cost models - you have to write and maintain both of them. And the optimizer needs to use when to use one versus when to use the other, and in cases where some row operators are wrapped, use both of them? It seems very complicated and a lot of development effort, no?
